File: puzzles/Einstein_Puzzle/C/einstein_puzzle.cpp
#include <stdio.h>
#include <string.h>
 
enum HouseStatus { Invalid, Underfull, Valid };
 
enum Attrib { C, M, D, A, S };
 
// Unfilled attributes are represented by -1
enum Colors { Red, Green, White, Yellow, Blue };
enum Mans { English, Swede, Dane, German, Norwegian };
enum Drinks { Tea, Coffee, Milk, Beer, Water };
enum Animals { Dog, Birds, Cats, Horse, Fish };
enum Smokes { PallMall, Dunhill, Blend, BlueMaster, Prince };
 
 
void printHouses(int ha[5][5]) {
    const char *color[] =  { "Red", "Green", "White", "Yellow", "Blue" };
    const char *man[] =    { "English", "Swede", "Dane", "German", "Norwegian" };
    const char *drink[] =  { "Tea", "Coffee", "Milk", "Beer", "Water" };
    const char *animal[] = { "Dog", "Birds", "Cats", "Horse", "Fish" };
    const char *smoke[] =  { "PallMall", "Dunhill", "Blend", "BlueMaster", "Prince" };
 
    printf("%-10.10s%-10.10s%-10.10s%-10.10s%-10.10s%-10.10s\n",
           "House", "Color", "Man", "Drink", "Animal", "Smoke");
 
    for (int i = 0; i < 5; i++) {
        printf("%-10d", i);
        if (ha[i][C] >= 0)
            printf("%-10.10s", color[ha[i][C]]);
        else
            printf("%-10.10s", "-");
        if (ha[i][M] >= 0)
            printf("%-10.10s", man[ha[i][M]]);
        else
            printf("%-10.10s", "-");
        if (ha[i][D] >= 0)
            printf("%-10.10s", drink[ha[i][D]]);
        else
            printf("%-10.10s", "-");
        if (ha[i][A] >= 0)
            printf("%-10.10s", animal[ha[i][A]]);
        else
            printf("%-10.10s", "-");
        if (ha[i][S] >= 0)
            printf("%-10.10s\n", smoke[ha[i][S]]);
        else
            printf("-\n");
    }
}
 
 
int checkHouses(int ha[5][5]) {
    int c_add = 0, c_or = 0;
    int m_add = 0, m_or = 0;
    int d_add = 0, d_or = 0;
    int a_add = 0, a_or = 0;
    int s_add = 0, s_or = 0;
 
    // Cond 9: In the middle house they drink milk.
    if (ha[2][D] >= 0 && ha[2][D] != Milk)
        return Invalid;
 
    // Cond 10: The Norwegian lives in the first house.
    if (ha[0][M] >= 0 && ha[0][M] != Norwegian)
        return Invalid;
 
    for (int i = 0; i < 5; i++) {
        // Uniqueness tests.
        if (ha[i][C] >= 0) {
            c_add += (1 << ha[i][C]);
            c_or |= (1 << ha[i][C]);
        }
        if (ha[i][M] >= 0) {
            m_add += (1 << ha[i][M]);
            m_or |= (1 << ha[i][M]);
        }
        if (ha[i][D] >= 0) {
            d_add += (1 << ha[i][D]);
            d_or |= (1 << ha[i][D]);
        }
        if (ha[i][A] >= 0) {
            a_add += (1 << ha[i][A]);
            a_or |= (1 << ha[i][A]);
        }
        if (ha[i][S] >= 0) {
            s_add += (1 << ha[i][S]);
            s_or |= (1 << ha[i][S]);
        }
 
        // Cond 2: The English man lives in the red house.
        if ((ha[i][M] >= 0 && ha[i][C] >= 0) &&
            ((ha[i][M] == English && ha[i][C] != Red) || // Checking both
             (ha[i][M] != English && ha[i][C] == Red)))  // to make things quicker.
            return Invalid;
 
        // Cond 3: The Swede has a dog.
        if ((ha[i][M] >= 0 && ha[i][A] >= 0) &&
            ((ha[i][M] == Swede && ha[i][A] != Dog) ||
             (ha[i][M] != Swede && ha[i][A] == Dog)))
            return Invalid;
 
        // Cond 4: The Dane drinks tea.
        if ((ha[i][M] >= 0 && ha[i][D] >= 0) &&
            ((ha[i][M] == Dane && ha[i][D] != Tea) ||
             (ha[i][M] != Dane && ha[i][D] == Tea)))
            return Invalid;
 
        // Cond 5: The green house is immediately to the left of the white house.
        if ((i > 0 && ha[i][C] >= 0 /*&& ha[i-1][C] >= 0 */ ) &&
            ((ha[i - 1][C] == Green && ha[i][C] != White) ||
             (ha[i - 1][C] != Green && ha[i][C] == White)))
            return Invalid;
 
        // Cond 6: drink coffee in the green house.
        if ((ha[i][C] >= 0 && ha[i][D] >= 0) &&
            ((ha[i][C] == Green && ha[i][D] != Coffee) ||
             (ha[i][C] != Green && ha[i][D] == Coffee)))
            return Invalid;
 
        // Cond 7: The man who smokes Pall Mall has birds.
        if ((ha[i][S] >= 0 && ha[i][A] >= 0) &&
            ((ha[i][S] == PallMall && ha[i][A] != Birds) ||
             (ha[i][S] != PallMall && ha[i][A] == Birds)))
            return Invalid;
 
        // Cond 8: In the yellow house they smoke Dunhill.
        if ((ha[i][S] >= 0 && ha[i][C] >= 0) &&
            ((ha[i][S] == Dunhill && ha[i][C] != Yellow) ||
             (ha[i][S] != Dunhill && ha[i][C] == Yellow)))
            return Invalid;
 
        // Cond 11: The man who smokes Blend lives in the house next to the house with cats.
        if (ha[i][S] == Blend) {
            if (i == 0 && ha[i + 1][A] >= 0 && ha[i + 1][A] != Cats)
                return Invalid;
            else if (i == 4 && ha[i - 1][A] != Cats)
                return Invalid;
            else if (ha[i + 1][A] >= 0 && ha[i + 1][A] != Cats && ha[i - 1][A] != Cats)
                return Invalid;
        }
 
        // Cond 12: In a house next to the house where they have a horse, they smoke Dunhill.
        if (ha[i][S] == Dunhill) {
            if (i == 0 && ha[i + 1][A] >= 0 && ha[i + 1][A] != Horse)
                return Invalid;
            else if (i == 4 && ha[i - 1][A] != Horse)
                return Invalid;
            else if (ha[i + 1][A] >= 0 && ha[i + 1][A] != Horse && ha[i - 1][A] != Horse)
                return Invalid;
        }
 
        // Cond 13: The man who smokes Blue Master drinks beer.
        if ((ha[i][S] >= 0 && ha[i][D] >= 0) &&
            ((ha[i][S] == BlueMaster && ha[i][D] != Beer) ||
             (ha[i][S] != BlueMaster && ha[i][D] == Beer)))
            return Invalid;
 
        // Cond 14: The German smokes Prince
        if ((ha[i][M] >= 0 && ha[i][S] >= 0) &&
            ((ha[i][M] == German && ha[i][S] != Prince) ||
             (ha[i][M] != German && ha[i][S] == Prince)))
            return Invalid;
 
        // Cond 15: The Norwegian lives next to the blue house.
        if (ha[i][M] == Norwegian &&
            ((i < 4 && ha[i + 1][C] >= 0 && ha[i + 1][C] != Blue) ||
             (i > 0 && ha[i - 1][C] != Blue)))
            return Invalid;
 
        // Cond 16: They drink water in a house next to the house where they smoke Blend.
        if (ha[i][S] == Blend) {
            if (i == 0 && ha[i + 1][D] >= 0 && ha[i + 1][D] != Water)
                return Invalid;
            else if (i == 4 && ha[i - 1][D] != Water)
                return Invalid;
            else if (ha[i + 1][D] >= 0 && ha[i + 1][D] != Water && ha[i - 1][D] != Water)
                return Invalid;
        }
 
    }
 
    if ((c_add != c_or) || (m_add != m_or) || (d_add != d_or)
        || (a_add != a_or) || (s_add != s_or)) {
        return Invalid;
    }
 
    if ((c_add != 0b11111) || (m_add != 0b11111) || (d_add != 0b11111)
        || (a_add != 0b11111) || (s_add != 0b11111)) {
        return Underfull;
    }
 
    return Valid;
}
 
 
int bruteFill(int ha[5][5], int hno, int attr) {
    int stat = checkHouses(ha);
    if ((stat == Valid) || (stat == Invalid))
        return stat;
 
    int hb[5][5];
    memcpy(hb, ha, sizeof(int) * 5 * 5);
    for (int i = 0; i < 5; i++) {
        hb[hno][attr] = i;
        stat = checkHouses(hb);
        if (stat != Invalid) {
            int nexthno, nextattr;
            if (attr < 4) {
                nextattr = attr + 1;
                nexthno = hno;
            } else {
                nextattr = 0;
                nexthno = hno + 1;
            }
 
            stat = bruteFill(hb, nexthno, nextattr);
            if (stat != Invalid) {
                memcpy(ha, hb, sizeof(int) * 5 * 5);
                return stat;
            }
        }
    }
 
    // We only come here if none of the attr values assigned were valid.
    return Invalid;
}
 
 
int main() {
    int ha[5][5] = {{-1, -1, -1, -1, -1}, {-1, -1, -1, -1, -1},
                    {-1, -1, -1, -1, -1}, {-1, -1, -1, -1, -1},
                    {-1, -1, -1, -1, -1}};
 
    bruteFill(ha, 0, 0);
    printHouses(ha);
 
    return 0;
}

File: puzzles/Einstein_Puzzle/Javascript/einstein_puzzle.js
const startTime = Date.now();
function generatePermutations(dictionary) {
  const results = [];
  const permute = (array, position) => {
    if (position === array.length - 1) {
      results.push(array);
    } else {
      for (let i = position; i < array.length; i++) {
        const swap = array[position];
        array.splice(position, 1, array[i]);
        array.splice(i, 1, swap);
        permute(array.slice(), (position + 1));
      }
    }
  };
  permute(dictionary, 0);
  return results;
};

const colors = generatePermutations(['red', 'white', 'green', 'yellow', 'blue']);
const drinks = generatePermutations(['tea', 'coffee', 'milk', 'beer', 'water']);
const smokes = generatePermutations(['Pall Mall', 'Dunhill', 'Blends', 'Bluemasters', 'Prince']);
const pets = generatePermutations(['dogs', 'birds', 'cats', 'fish', 'horse']);
const nationalities = generatePermutations(['Brit', 'Swede', 'Dane', 'Norwegian', 'German']);

function print(color, drink, smoke, pet, nationality) {
  console.log('------');
  [0, 1, 2, 3, 4].forEach(index => {
    console.log('House %s: The %s lives in a %s house, drinks %s, smokes %s, and has %s.',
      index + 1, nationality[index], color[index], drink[index], smoke[index], pet[index]);
  })
  console.log('------');
}

const reasons = {};


let n = 0;
function inc(reason) {
  reasons[reason] = (reasons[reason] || 0) + 1;
}


nationalities.forEach(nationality => {
  // 9. The Norwegian lives in the first house.
  if (nationality[0] !== 'Norwegian') {
    return inc('9. The Norwegian lives in the first house.');
  }
  smokes.forEach(smoke => {
    pets.forEach(pet => {
      // 11. The owner who keeps the horse lives next to the one who smokes Dunhill.
      if (smoke[pet.indexOf('horse') - 1] !== 'Dunhill' && smoke[pet.indexOf('horse') + 1] !== 'Dunhill') {
        return inc('11. The owner who keeps the horse lives next to the one who smokes Dunhill.');
      }
      // 2. The Swede keeps dogs as pets.
      if (pet[nationality.indexOf('Swede')] !== 'dogs') {
        return inc('2. The Swede keeps dogs as pets.');
      }
      // 10. The owner who smokes Blends lives next to the one who keeps cats.
      if (pet[smoke.indexOf('Blends') - 1] !== 'cats' && pet[smoke.indexOf('Blends') + 1] !== 'cats') {
        return inc('10. The owner who smokes Blends lives next to the one who keeps cats.');
      }
      // 6. The owner who smokes Pall Mall rears birds.
      if (pet[smoke.indexOf('Pall Mall')] !== 'birds') {
        return inc('6. The owner who smokes Pall Mall rears birds.');
      }
      drinks.forEach(drink => {
        // 8. The owner living in the center house drinks milk.
        if (drink[2] !== 'milk') {
          return inc('8. The owner living in the center house drinks milk.');
        }
        colors.forEach(color => {
          // 1. The Brit lives in the red house.
          if (color[nationality.indexOf('Brit')] !== 'red') {
            return inc('1. The Brit lives in the red house.');
          }
          // 3. The Dane drinks tea.
          if (drink[nationality.indexOf('Dane')] !== 'tea') {
            return inc('3. The Dane drinks tea.');
          }
          // 4. The green house is on the immediate left of the white house.
          if (color[color.indexOf('green') + 1] !== 'white') {
            return inc('4. The green house is on the immediate left of the white house.');
          }
          // 5. The green house's owner drinks coffee.
          if (drink[color.indexOf('green')] !== 'coffee') {
            return inc('5. The green house\'s owner drinks coffee.');
          }
          // 7. The owner of the yellow house smokes Dunhill.
          if (smoke[color.indexOf('yellow')] !== 'Dunhill') {
            return inc('7. The owner of the yellow house smokes Dunhill.');
          }
          // 12. The owner who smokes Bluemasters drinks beer.
          if (drink[smoke.indexOf('Bluemasters')] !== 'beer') {
            return inc('12. The owner who smokes Bluemasters drinks beer.');
          }
          // 13. The German smokes Prince.
          if (smoke[nationality.indexOf('German')] !== 'Prince') {
            return inc('13. The German smokes Prince.');
          }
          // 14. The Norwegian lives next to the blue house.
          if (color[nationality.indexOf('Norwegian') - 1] !== 'blue' &&
            color[nationality.indexOf('Norwegian') + 1] !== 'blue') {
            return inc('14. The Norwegian lives next to the blue house.');
          }
          // 15. The owner who smokes Blends lives next to the one who drinks water.
          if (drink[smoke.indexOf('Blends') - 1] !== 'water' &&
            drink[smoke.indexOf('Blends') + 1] !== 'water') {
            return inc('15. The owner who smokes Blends lives next to the one who drinks water.');
          }

          print(color, drink, smoke, pet, nationality);
        });
      });
    });
  });
});

// Print reasons
let total = 0;
const totalTime = Date.now() - startTime;
Object
  .keys(reasons)
  .sort((a, b) => reasons[b] - reasons[a]) // reverse order (highest first)
  .forEach(reason => {
    total += reasons[reason];
    console.log(reasons[reason], reason);
  });
console.log('--------\n%d total nodes tested. (%d%% of total space)  %s seconds', total, (100 * total / Math.pow(120, 5)).toPrecision(3), totalTime / 1000);
